File: api/Sources/Entity/Diary.swift
import Foundation

public struct Diary: Hashable, Equatable {
    
    public var 日付: Date
    
    public var 天候: Weather
    
    public var 日直者: [Student]
    
    public var 授業: [Study]
    
    public var 行時・その他: String
    
    public var 清掃担当研究室: [Research]
    
    public var 清掃実施（予定）日: Date
    
    public var 感想・連絡事項等: String
    
    public init(
        日付: Date,
        天候: Weather,
        日直者: [Student],
        授業: [Study],
        行時・その他: String,
        清掃担当研究室: [Research],
        清掃実施（予定日）: Date,
        感想・連絡事項等: String
    ) {
        self.日付 = 日付
        self.天候 = 天候
        self.日直者 = 日直者
        self.授業 = 授業
        self.行時・その他 = 行時・その他
        self.清掃担当研究室 = 清掃担当研究室
        self.清掃実施（予定）日 = 清掃実施（予定日）
        self.感想・連絡事項等 = 感想・連絡事項等
    }
}

File: app/src/entity/diary/diary.ts
import Student from '@/entity/diary/student/student';
import Study from '@/entity/diary/study/study';
import Research from '@/entity/diary/research/research';
import { z } from 'zod';
import { StudentSchema } from '@/entity/diary/student/student';
import { StudySchema } from '@/entity/diary/study/study';
import { ResearchSchema } from '@/entity/diary/research/research';
import type { Decoder } from '@/resources/codable/decode';
import { createArrayDecoder } from '@/resources/codable/decode';
import type Encodable from '@/resources/codable/encode';
import encodeUTCDate from '@/resources/codable/date';
import { encodeArray } from '@/resources/codable/encode';

export type Weather = '晴れ' | '雨' | '曇り' | '雪';
export const WeatherSchema = z.enum(['晴れ', '雨', '曇り', '雪']);

export default class Diary implements Encodable {
	/**
	 * 日誌
	 * @param date 日付
	 * @param weather 天候
	 * @param responsibleStudents 日直担当
	 * @param studies 授業一覧
	 * @param eventOrOther 行事・その他
	 * @param cleanResponsibleResearches 清掃担当研究室
	 * @param cleanDate 清掃（予定）日
	 * @param impressionOrContact 感想・連絡事項
	 */
	constructor(
		readonly date: Date,
		readonly weather: Weather,
		readonly responsibleStudents: readonly Student[],
		readonly studies: readonly Study[],
		readonly eventOrOther: string,
		readonly cleanResponsibleResearches: readonly Research[],
		readonly cleanDate: Date,
		readonly impressionOrContact: string
	) {}

	static readonly decode: Decoder<Diary> = (json) => {
		const schema = DiarySchema.parse(json);
		return new Diary(
			new Date(schema.date),
			schema.weather,
			createArrayDecoder(Student.decode)(schema.responsibleStudents),
			createArrayDecoder(Study.decode)(schema.studies),
			schema.eventOrOther,
			createArrayDecoder(Research.decode)(schema.cleanResponsibleResearches),
			new Date(schema.cleanDate),
			schema.impressionOrContact
		);
	};

	encode(): unknown {
		return {
			date: encodeUTCDate(this.date),
			weather: this.weather,
			responsibleStudents: encodeArray(this.responsibleStudents),
			studies: encodeArray(this.studies),
			eventOrOther: this.eventOrOther,
			cleanResponsibleResearches: encodeArray(this.cleanResponsibleResearches),
			cleanDate: encodeUTCDate(this.cleanDate),
			impressionOrContact: this.impressionOrContact
		};
	}
}

export const DiarySchema = z.object({
	date: z.string().datetime(),
	weather: WeatherSchema,
	responsibleStudents: z.array(StudentSchema),
	studies: z.array(StudySchema),
	eventOrOther: z.string(),
	cleanResponsibleResearches: z.array(ResearchSchema),
	cleanDate: z.string().datetime(),
	impressionOrContact: z.string()
});
